src/lib.rs

```diff
@@ -11,31 +12,64 @@ use rustix::mm::{mmap_anonymous, munmap, MapFlags, ProtFlags};
 
 struct BufRingMmap {
     ptr: NonNull<BufRingEntry>,
+    len: usize,
     size: usize,
 }
 
 impl BufRingMmap {
-    fn new(size: usize) -> io::Result<Self> {
+    fn new(len: usize) -> io::Result<Self> {
+        let size = len * std::mem::size_of::<BufRingEntry>();
+        // SAFETY: correctly aligned.
         let ptr = unsafe {
-            // Safety: we will check the ptr
             mmap_anonymous(
                 ptr::null_mut(),
                 size,
                 ProtFlags::READ | ProtFlags::WRITE,
                 MapFlags::PRIVATE,
             )
         }?;
-        let ptr = NonNull::new(ptr)
-            .ok_or_else(|| io::Error::new(ErrorKind::Other, "mmap return null ptr"))?;
+        // SAFETY: rustix checkes the pointer.
+        let ptr = unsafe { NonNull::new_unchecked(ptr.cast()) };
 
-        Ok(Self {
-            ptr: ptr.cast(),
-            size,
-        })
+        Ok(Self { ptr, len, size })
+    }
+
+    fn as_slice_uninit_mut(&mut self) -> &mut [MaybeUninit<BufRingEntry>] {
+        // SAFETY: the pointer is valid
+        unsafe { std::slice::from_raw_parts_mut(self.ptr.as_ptr().cast(), self.len) }
+    }
+
+    fn atomic_tail(&self) -> &AtomicU16 {
+        // Safety: no one read/write tail ptr without atomic operation after init
+        unsafe { AtomicU16::from_ptr(BufRingEntry::tail(self.ptr.as_ptr()).cast_mut()) }
+    }
+
+    /// # Safety
+    ///
+    /// Caller must ensure the ring is not full.
+    unsafe fn add_buffer(
+        &mut self,
+        buf: &mut [MaybeUninit<u8>],
+        bid: u16,
+        mask: u16,
+        buf_offset: usize,
+    ) {
+        let tail = self.atomic_tail();
+        let index = ((tail.load(Ordering::Acquire) + (buf_offset as u16)) & mask) as usize;
+
+        // SAFETY: only write plain data here
+        let buf_ring_entry = (&mut self.as_slice_uninit_mut()[index]).assume_init_mut();
+
+        buf_ring_entry.set_addr(buf.as_ptr() as _);
+        buf_ring_entry.set_len(buf.len() as _);
+        buf_ring_entry.set_bid(bid);
     }
 
-    fn as_ptr(&self) -> *mut BufRingEntry {
-        self.ptr.as_ptr()
+    /// # Safety
+    ///
+    /// The advanced count should be equal to the added count.
+    unsafe fn advance_buffers(&self, count: u16) {
+        self.atomic_tail().fetch_add(count, Ordering::Release);
```

Comment:
Do we have to `& mask` here?

Reply:
https://github.com/axboe/liburing/blob/7b3245583069bd481190c9da18f22e9fc8c3a805/src/include/liburing.h#L1496

`fetch_add` is enough